arangod/RocksDBEngine/RocksDBIncrementalSync.cpp

```diff
@@ -44,6 +46,107 @@
 
 namespace arangodb {
+
+// remove all keys that are below first remote key or beyond last remote key
+Result removeKeysOutsideRange(VPackSlice chunkSlice, 
+                              LogicalCollection* col,
+                              OperationOptions& options,
+                              InitialSyncerIncrementalSyncStats& stats) {
+  size_t const numChunks = chunkSlice.length();
+
+  if (numChunks == 0) {
+    // no need to do anything
+    return Result();
+  }
+  
+  // first chunk
+  SingleCollectionTransaction trx(
+    transaction::StandaloneContext::Create(col->vocbase()),
+    *col,
+    AccessMode::Type::EXCLUSIVE
+  );
+
+  trx.addHint(
+      transaction::Hints::Hint::RECOVERY);  // to turn off waitForSync!
+  trx.addHint(transaction::Hints::Hint::NO_TRACKING);
+  trx.addHint(transaction::Hints::Hint::NO_INDEXING);
+
+  Result res = trx.begin();
+
+  if (!res.ok()) {
+    return Result(
+        res.errorNumber(),
+        std::string("unable to start transaction: ") + res.errorMessage());
+  }
+
+  VPackSlice chunk = chunkSlice.at(0);
+
+  TRI_ASSERT(chunk.isObject());
+  auto lowSlice = chunk.get("low");
+  TRI_ASSERT(lowSlice.isString());
+  StringRef lowRef(lowSlice);
+
+  // last high
+  chunk = chunkSlice.at(numChunks - 1);
+  TRI_ASSERT(chunk.isObject());
+
+  auto highSlice = chunk.get("high");
+  TRI_ASSERT(highSlice.isString());
+  StringRef highRef(highSlice);
+
+  LogicalCollection* coll = trx.documentCollection();
+  auto iterator = createPrimaryIndexIterator(&trx, coll);
+
+  VPackBuilder builder;
+
+  // remove everything from the beginning of the key range until the lowest
+  // remote key
+  iterator.next(
+      [&](rocksdb::Slice const& rocksKey, rocksdb::Slice const& rocksValue) {
+        StringRef docKey(RocksDBKey::primaryKey(rocksKey));
+        if (docKey.compare(lowRef) < 0) {
+          builder.clear();
+          builder.add(velocypack::ValuePair(docKey.data(), docKey.size(),
+                                            velocypack::ValueType::String));
+          trx.remove(col->name(), builder.slice(), options);
+          ++stats.numDocsRemoved;
+          // continue iteration
+          return true;
+        }
+
+        // stop iteration
+        return false;
+      },
+      std::numeric_limits<std::uint64_t>::max());
+  
+  // remove everything from the highest remote key until the end of the key range
+  auto index = col->lookupIndex(0); //RocksDBCollection->primaryIndex() is private
+  TRI_ASSERT(index->type() == Index::IndexType::TRI_IDX_TYPE_PRIMARY_INDEX);
+  auto primaryIndex = static_cast<RocksDBPrimaryIndex*>(index.get());
+
+  RocksDBKeyLeaser key(&trx);
+  key->constructPrimaryIndexValue(primaryIndex->objectId(), highRef);
+  iterator.seek(key->string());
+
+  iterator.next(
+      [&](rocksdb::Slice const& rocksKey, rocksdb::Slice const& rocksValue) {
+        StringRef docKey(RocksDBKey::primaryKey(rocksKey));
+        if (docKey.compare(highRef) > 0) {
```

Comment:
couldn't you merge these callbacks ?
Just call it once, then do a seek to `highRef` and call next with the callback again.